### ISSUE TITLE
Spec and Status of CPU are different style in HPA

### ISSUE THREAD

**Issue by erictune** [kubernetes/kubernetes#18527]
Kind `HorizontalPodAutoscaler` has field `spec.cpuUtilization.targetPercentage`, but `status. currentCPUUtilizationPercentage`.

Why not the same?  For example:
 
-  `spec.targetCPUUtilizationPercentage` and `status. currentCPUUtilizationPercentage`, or
-  `spec.cpuUtilization.targetPercentage` and `status.cpuUtilization.currentPercentage`.

**Comment by erictune** [comment]
@jszczepkowski @mwielgus @bgrant0607

**Comment by jszczepkowski** [comment]
Thanks for rising this! I agree that current solution is nonuniform. I propose be another way of dealing with it. In spec: CPUTargetUtilization struct with TargetPercentage inside. In status: CPUCurrentValue struct with CurrentPercentage inside. So, we will have two different structs for status and spec. Having two different struct will give us flexibility in future.

I'll be happy to clean it up after we made decision which way to go.

**Comment by jszczepkowski** [comment]
CC @DirectXMan12 @ncdc

**Comment by DirectXMan12** [comment]
> CPUTargetUtilization struct with TargetPercentage inside. In status: CPUCurrentValue struct with CurrentPercentage inside.

Having target/current both in the struct name and in the field name seems a bit redundant.  Not a blocker, per se, but I do like "spec.cpuUtilization.targetPercentage and status.cpuUtilization.currentPercentage" a bit better.

**Comment by jszczepkowski** [comment]
> Having target/current both in the struct name and in the field name seems a bit redundant. Not a blocker, per se, but I do like "spec.cpuUtilization.targetPercentage and status.cpuUtilization.currentPercentage" a bit better.

What will be the type of cpuUtilization than? I guess we need different types for spec and status. If we will have different types for spec and status: I agree.

**Comment by DirectXMan12** [comment]
Yeah, they'd both have a field name of "cpuUtilization" (or at least have their serialized field name be "cpuUtiliziation"), but have different struct types.

**Comment by ncdc** [comment]
If we do make a change, please keep things backwards compatible and provide a conversion to continue supporting the old field.

**Comment by bgrant0607** [comment]
I don't remember the reason for the extra level of struct in spec. It looks vestigial:
https://github.com/kubernetes/kubernetes/pull/15706/files

Not sure if we might want to report status in multiple formats in the future, though, for which the nested structure would be better.

**Comment by erictune** [comment]
@ncdc are you asking for an automatic conversion from `extensions/v1beta1/horizontalpodautoscaler` to `extensions/v1/horizontalpodautoscaler`?  That doesn't help if we change status?  Or are you asking for the new status object to have the same info in two places?

**Comment by ncdc** [comment]
@erictune I want to make sure we retain backwards compatibility when using `extensions/v1beta1/horizontalpodautoscaler`; i.e., don't rename fields, don't delete fields.

**Comment by piosz** [comment]
So there are 3 options here:
 1. remove extra level of struct in spec if there is no reason for having it
 2. make the fields to follow the same pattern (and decide which one)
 3. leave without any change

Personally I prefer option 1 as it is simple and I can't see any use case to have nested struct in spec.

**Comment by ncdc** [comment]
@piosz just to reiterate, please don't make any backwards-incompatible changes to `extensions/v1beta1/horizontalpodautoscaler` - thanks!

**Comment by piosz** [comment]
@ncdc IIUC the only possible option is no. 3 then. Am I right?

**Comment by ncdc** [comment]
@piosz we could add another field to v1beta1 and then have 2 copies of it, but that gets nasty - how do you resolve conflicts if both are updated to different values? I think it makes sense to do option 1 for v1, but not v1beta1.

**Comment by piosz** [comment]
I'm planning to graduate HPA API to GA for 1.2 (see #18528), so I'll move related objects from v1beta1 to v1. IIUC you are ok with changing the fields then?

**Comment by ncdc** [comment]
Yes, I'm fine changing things in v1 as long as we can convert between v1beta1 and v1 successfully.

**Comment by piosz** [comment]
Ack

**Comment by piosz** [comment]
So the final proposal is to change `spec.cpuUtilization.targetPercentage` to `spec.targetCPUUtilizationPercentage` in v1.

**Comment by liggitt** [comment]
> I'll move related objects from v1beta1 to v1. IIUC you are ok with changing the fields then?

to be clear, you mean copy the related objects from v1beta1 to v1 and add conversions?

**Comment by piosz** [comment]
exactly

**Comment by fgrzadkowski** [comment]
@piosz Is this fixed by #20501?

**Comment by fgrzadkowski** [comment]
We now have the following fields:
```
	// target average CPU utilization (represented as a percentage of requested CPU) over all the pods;
	TargetCPUUtilizationPercentage *int32 `json:"targetCPUUtilizationPercentage,omitempty"`
```
and
```
	// current average CPU utilization over all pods, represented as a percentage of requested CPU,
	// e.g. 70 means that an average pod is using now 70% of its requested CPU.
	CurrentCPUUtilizationPercentage *int32 `json:"currentCPUUtilizationPercentage,omitempty"`
```
I believe this is unified. @piosz please re-open if I'm mistaken.

**Comment by piosz** [comment]
I can confirm that it's fixed by #20501
